### PR TITLE
[WIP] Add UA_DATATYPEKIND_NULL to align with Variant types

### PR DESCRIPTION
**Pull request by electro-logic**
This PR add UA_DATATYPEKIND_NULL to align UA_DATATYPEKIND with the Variant types (https://reference.opcfoundation.org/Core/Part6/v105/docs/5.1.2#_Ref83387521) and is related to issue #6780 

Still needs some help to fix automatically generated types.

types_generated.h contains types starting from 0, while the addition of the UA_DATATYPEKIND_NULL should now make them start from 1.

```
#define UA_TYPES_COUNT 388
extern UA_EXPORT UA_DataType UA_TYPES[UA_TYPES_COUNT];

/* Boolean */
#define UA_TYPES_BOOLEAN 0

/* SByte */
#define UA_TYPES_SBYTE 1

```

should become

```
#define UA_TYPES_COUNT 389
extern UA_EXPORT UA_DataType UA_TYPES[UA_TYPES_COUNT];

/* Undefined type */
#define UA_TYPES_NULL 0

/* Boolean */
#define UA_TYPES_BOOLEAN 1

/* SByte */
#define UA_TYPES_SBYTE 2

```

some changes to generate_datatypes.py are probably required, any help is welcome.

This PR is also related to https://github.com/open62541/open62541/issues/6860

### CODE REVIEW

**Comment by jpfr** [comment]
The data type kind is a different index from the UA_TYPES array.

the kind index is used mostly internally for jump tables.

You probably can simplify quite a bit of the proposed changes.

**Comment by electro-logic** [comment]
Hi, thank you for the reply.

The issue is that a lot of code (library and user code) is often retrieving the UA_DATATYPE by doing UA_TYPES[kind] where kind is of type UA_DATATYPEKIND.

I think that would be much better to keep the internal values aligned. A simple way to align them would be to add someway a "virtual" hardcoded UA_TYPES_NULL node. What do you think about this direction?

**Comment by jpfr** [comment]
Lookup of the type via the kind only works in some cases. Because one kind is “structure” which maps to many types.
The pattern of looking up the type via the kind should be avoided!!

How would the PR improve the situation?
Because you increase both the kind and the normal array index.

**Comment by electro-logic** [comment]
> Lookup of the type via the kind only works in some cases. Because one kind is “structure” which maps to many types. The pattern of looking up the type via the kind should be avoided!!

Yes, the conversion is done only when the kind value is <= UA_DATATYPEKIND_DIAGNOSTICINFO, in a similar way to ENCODE_BINARY(Variant) / DECODE_BINARY(Variant)

> How would the PR improve the situation? Because you increase both the kind and the normal array index.

Aligning with the variant type avoid the +/-1 conversion in ENCODE_BINARY(Variant) / DECODE_BINARY(Variant) and makes the code cleaner. Is also safer because makes clear when UA_DATATYPEKIND is not initialized.

File: src/ua_types.c [hunk 1]
```diff
@@ -1334,6 +1334,7 @@ copyNotImplemented(const void *src, void *dst, const UA_DataType *type) {
 }
 
 const UA_copySignature copyJumpTable[UA_DATATYPEKINDS] = {
+    NULL,
```
Comment (jpfr) [inline src/ua_types.c:1337]: This should point to a function that returns an error…

File: src/ua_types.c [hunk 2]
```diff
@@ -1437,6 +1438,7 @@ static void nopClear(void *p, const UA_DataType *type) { }
 
 const
 UA_clearSignature clearJumpTable[UA_DATATYPEKINDS] = {
+    NULL,
```
Comment (jpfr) [inline src/ua_types.c:1441]: The same here.

also for binary, JSON and xml encoding.

**Comment by electro-logic** [comment]
Note: I'm still trying to optimize my dev environment for open62541, but unit tests are not working under Windows / Visual Studio 2022 (see https://github.com/open62541/open62541/issues/6859). Some PR may not pass basic checks because of this.

File: src/ua_types.c (hunk 2 above)
Comment (electro-logic) [reply src/ua_types.c:1441]: good point, for now I put a NULL to compile and trying to test the happy path as a first step.

**Comment by jpfr** [comment]
If the only advantage is that we can save a +1 or -1 on the indices in the binary decoding, then this is not worth the effort.

I agree it would be an improvement.
But there is just too much risk of either breaking existing user code or overlooking an edge case, leading to bugs.

**Comment by electro-logic** [comment]
Let's re-evaluate this PR for a breaking changes release (ex. v2.0).

**Comment by jpfr** [comment]
:+1

**Pull request closed without merging**